python/mpt/src/mpt/trie_diff.cairo

```diff
@@ -1430,59 +1579,102 @@ func _compute_left_branch_on_right_extension_node{
     main_trie_end: AddressAccountDiffEntry*,
     storage_tries_end: StorageDiffEntry*,
 }(
-    left: BranchNode,
-    right: ExtensionNode,
+    subnodes: Subnodes,
+    right: OptionalInternalNode,
+    parent_left: OptionalInternalNode,
+    parent_right: OptionalInternalNode,
     path: Bytes,
     account_address: OptionalAddress,
     index: felt,
 ) -> () {
     alloc_locals;
-
     if (index == 16) {
         return ();
     }
 
-    let subnodes_ptr = cast(left.value.subnodes.value, felt*);
+    // Use `branch_0` as the base pointer to the list of subnodes and index it as a felt*.
+    let subnodes_ptr = cast(subnodes.value, felt*);
     let subnode_i_extended = Extended(cast(subnodes_ptr[index], ExtendedEnum*));
     let subnode_i = OptionalUnionInternalNodeExtendedImpl.from_extended(subnode_i_extended);
 
+    // TODO: optimization possible here
+
     // path = path + bytes([i])
     let path_copy = Bytes__copy__(path);
     assert path_copy.value.data[path_copy.value.len] = index;
     tempvar sub_path = Bytes(
         new BytesStruct(data=path_copy.value.data, len=path_copy.value.len + 1)
     );
 
-    let first_nib = right.value.key_segment.value.data[0];
+    // Two cases:
+    // 1. The first nibble of the subnode matches the branch node's key segment, in which
+    // case we compare to the subnode
+    // 2. It doesn't in which case we compare to null.
+
+    let right_extension = right.value.extension_node;
+    let first_nib = right_extension.value.key_segment.value.data[0];
     if (first_nib == index) {
-        // Compare to the shortened extension node
-        // the length of the key segment is always _at least_ one in an extension
-        tempvar extension = ExtensionNode(
-            new ExtensionNodeStruct(
-                key_segment=Bytes(
-                    new BytesStruct(
-                        data=right.value.key_segment.value.data + 1,
-                        len=right.value.key_segment.value.len - 1,
+        // Fully consumed by this nibble: compare to the subnode
+        if (right_extension.value.key_segment.value.len == 1) {
+            let node_to_compare_ = OptionalUnionInternalNodeExtendedImpl.from_extended(
+                right_extension.value.subnode
+            );
+            tempvar current_parent_right = right;
+            tempvar node_to_compare = node_to_compare_;
```

Comment:
what would happen if we didn't do this?

Reply:
if you mean:
- checking whether the length of the key segment is exactly one? if we didn't we could shorten an extension node to have a keysegment of length 0 and that might be a problem
- if the keysegment len >= 1, we need to compare the branch subnode to the shortened extension, if key segment len == 1 we must compare the branch subnode to the extension subnode